### PR TITLE
Refine Windows performance builds

### PR DESCRIPTION
**Pull request by blindpirate**
1. Use `subst` to avoid long path issue.
2. Introduce Os.capitalized() and Os.lowerCase().
3. Refactoring.

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle test this

File: .teamcity/common/Os.kt [hunk 1]
```diff
@@ -16,22 +16,22 @@
...
 package common
 
-val killAllGradleProcessesUnixLike = """
-    free -m
-    ps aux | egrep 'Gradle(Daemon|Worker)'
-    ps aux | egrep 'Gradle(Daemon|Worker)' | awk '{print ${'$'}2}' | xargs kill -9
-    free -m
-    ps aux | egrep 'Gradle(Daemon|Worker)' | awk '{print ${'$'}2}'
-""".trimIndent()
+const val killAllGradleProcessesUnixLike = """
```
Comment (wolfs) [inline .teamcity/common/Os.kt:19]: Can this be `private` as well?

File: .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt [hunk 2]
```diff
@@ -43,14 +38,12 @@ class IndividualPerformanceScenarioWorkers(model: CIBuildModel, os: Os = Os.LINU
...
-            name = "KILL_GRADLE_PROCESSES"
-            executionMode = BuildStep.ExecutionMode.ALWAYS
-            scriptContent = os.killAllGradleProcesses
-        }
+        killGradleProcessesStep(os)
+        windowsSubstrStep(os)
         gradleWrapper {
             name = "GRADLE_RUNNER"
             tasks = ""
+            workingDir = "P:/"
```
Comment (wolfs) [inline .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt:46]: The workingDir should only be `P:\` when on Windows.

File: .teamcity/common/performance-test-extensions.kt [hunk 3]
```diff
@@ -55,8 +58,37 @@ fun distributedPerformanceTestParameters(workerId: String = "Gradle_Check_Indivi
...
+    if (os == Os.WINDOWS) {
+        script {
+            name = "SETUP_VIRTUAL_DISK_FOR_PERF_TEST"
+            executionMode = BuildStep.ExecutionMode.ALWAYS
+            scriptContent = """subst p: "%teamcity.build.checkoutDir%" """
+        }
+    }
+}
+
+fun BuildSteps.removeWindowsSubstrStep(os: Os) {
```
Comment (wolfs) [inline .teamcity/common/performance-test-extensions.kt:86]:
```suggestion
fun BuildSteps.removeWindowsSubstStep(os: Os) {
```

This is `subst` not `substr`.

File: .teamcity/common/performance-test-extensions.kt [hunk 4]
```diff
@@ -55,8 +58,37 @@ fun distributedPerformanceTestParameters(workerId: String = "Gradle_Check_Indivi
...
+fun BuildSteps.killGradleProcessesStep(os: Os) {
+    script {
+        name = "KILL_GRADLE_PROCESSES"
+        executionMode = BuildStep.ExecutionMode.ALWAYS
+        scriptContent = os.killAllGradleProcesses
+    }
+}
+
+// to avoid pathname too long error
+fun BuildSteps.windowsSubstrStep(os: Os) {
```
Comment (wolfs) [inline .teamcity/common/performance-test-extensions.kt:76]: A better name for the methods could be `substDirOnWindows` and `removeSubstDirOnWindows`.

File: .teamcity/Gradle_Check/configurations/FunctionalTest.kt [hunk 5]
```diff
@@ -45,7 +45,7 @@ class FunctionalTest(
 
     applyTestDefaults(model, this, testTasks, notQuick = !testCoverage.isQuick, os = testCoverage.os,
         extraParameters = (
-            listOf(""""-PtestJavaHome=%${testCoverage.os.name.toLowerCase()}.${testCoverage.testJvmVersion}.${testCoverage.vendor}.64bit%"""") +
+            listOf(""""-PtestJavaHome=%${testCoverage.os}.${testCoverage.testJvmVersion}.${testCoverage.vendor}.64bit%"""") +
```
Comment (wolfs) [inline .teamcity/Gradle_Check/configurations/FunctionalTest.kt:48]:
```suggestion
            listOf(""""-PtestJavaHome=%${testCoverage.os.lowerCase()}.${testCoverage.testJvmVersion}.${testCoverage.vendor}.64bit%"""") +
```

Shouldn't this be ☝️ ?

File: .teamcity/common/Os.kt [hunk 6]
```diff
@@ -43,4 +43,8 @@ enum class Os(val agentRequirement: String, val ignoredSubprojects: List<String>
     );
 
     fun escapeKeyValuePair(key: String, value: String) = if (this == WINDOWS) """$key="$value"""" else """"$key=$value""""
+
+    fun lowerCase() = name.toLowerCase()
```
Comment (wolfs) [inline .teamcity/common/Os.kt:47]: I think these methods shouldn't say what they do but what purpose the string the produce have:

I think we need some version of the `Os.name` for
- Displaying it in the UI (-> `toString()`)
- Use as part of Teamcity IDs (-> `asId` or something like that, depending on what we use in other places)
- As part of the Teamcity property for the Java home. Maybe that could go to another class/method or be something like `toJavaHomePropertyId()`.

WDYT?

**Review by wolfs — CHANGES_REQUESTED** [review]

**Comment by blindpirate** [comment]
@wolfs PATL!

File: .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt [hunk 7]
```diff
@@ -14,9 +14,9 @@ import jetbrains.buildServer.configs.kotlin.v2019_2.AbsoluteId
 import model.CIBuildModel
 
 class IndividualPerformanceScenarioWorkers(model: CIBuildModel, os: Os = Os.LINUX) : BaseGradleBuildType(model, init = {
-    uuid = model.projectPrefix + "IndividualPerformanceScenarioWorkers${os.name.toLowerCase().capitalize()}"
+    uuid = model.projectPrefix + "IndividualPerformanceScenarioWorkers$os"
```
Comment (wolfs) [inline .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt:17]: I am not convinced we should use the `toString()` representation for the ID, even though it is the same string now.

**Review by wolfs — CHANGES_REQUESTED** [review]

File: .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt (hunk 7 above)
Comment (blindpirate) [reply .teamcity/Gradle_Check/configurations/IndividualPerformanceScenarioWorkers.kt:17]: How about `asName` or `asTypeId`?

**Comment by blindpirate** [comment]
@bot-gradle test this

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by blindpirate** [comment]
Thanks @wolfs !

**Pull request merged**